repair/repair.cc

```diff
@@ -2135,6 +2136,8 @@ future<> repair_service::repair_tablets(repair_uniq_id rid, sstring keyspace_nam
             } else if (data_centers.size() > 0 || hosts.size() > 0) {
                 throw std::runtime_error("You need to run primary range repair on all nodes in the cluster.");
             }
+            primary_replicas = co_await tmap.calc_primary_replica_map(tid, erm->get_topology(),
+                    select_primary_ranges_within_dc ? mydc : "");
```

Comment:
This is very wasteful. This shard (or node?) holds just a few tablets, yet you run a calculation that scans all the tablets for the table.

Reply:
primary selection is per node, not per shard.
I agree we can skip the selection for the repair use case if there's no replica on the node anyhow.
I'm not sure if that holds true also for the storage_service api case as well.
It might need to know primary ownership on other nodes as well for simulating the vnodes ownership api.

Reply:
What if there's >=1 tablet replica per node?

You're doing O(cluster) work on every node.

Reply:
> What if there's >=1 tablet replica per node?

Of the same tablet?  That would be a bug.

> You're doing O(cluster) work on every node.

If we persist the primary selection in the table we could calculate it once and reevaluate it on split / merge and maybe migrations.

Reply:
> > What if there's >=1 tablet replica per node?
> 
> Of the same tablet? That would be a bug.

No, I mean every node holds at least one tablet replica. This is expected to be the common case and negates the optimization ("we can skip the selection for the repair use case if there's no replica on the node anyhow").

> 
> > You're doing O(cluster) work on every node.
> 
> If we persist the primary selection in the tablets topology we could calculate it once and reevaluate it on split / merge and maybe migrations.

It should be persisted, and part of the main data structure, not yet another copy.

Reply:
Or calculated with the modulo algorithm, I don't know why you abandoned it.

Reply:
> > > What if there's >=1 tablet replica per node?
> > 
> > 
> > Of the same tablet? That would be a bug.
> 
> No, I mean every node holds at least one tablet replica. This is expected to be the common case and negates the optimization ("we can skip the selection for the repair use case if there's no replica on the node anyhow").

It would be the common case when RF==|nodes in dc| which might be a common case, but larger installations would have RF < |nodes in dc| and may benefit from the optimization

> > > You're doing O(cluster) work on every node.
> > 
> > 
> > If we persist the primary selection in the tablets topology we could calculate it once and reevaluate it on split / merge and maybe migrations.
> 
> It should be persisted, and part of the main data structure, not yet another copy.

Let's keep in mind that we should select a global primary replica as well as primary per DC.
I wonder if it would be useful in the future if raft groups will be based on the tablet replica where the global and per-dc primary replicas could be used for electing the group global leader and possibly per-dc gateway/coordinator.

> Or calculated with the modulo algorithm, I don't know why you abandoned it.

I abandoned it because of https://github.com/scylladb/scylladb/pull/18784#discussion_r1615262160
I think that your observation is correct that the modulo-based primary selection algorithm is susceptible to bad distribution of replicas.  Therefore I introduced the load_map algorithm that tries to balance the primary-replica load evenly, by providing a counter-bias to the bad distribution.

I'm ok with going back to that for the short term and continue working on persisting the primary ownership as a follow up for the longer term.

Reply:
Thinking of LWT, the primary replica has a special role of being the write leader/coordinator, be it paxos or raft. The information is relevant to the clients as well: Scylla's shard-aware drivers direct LWT writes to primaries if they know they are executing a conditional statement.

I'm not sure therefore it's correct to distribute the primaries across based on the load (i.e. number of tablets per node), some nodes may get more heat than others this way. But mainly, the clients should be able to know which tablet replica owner is the primary one, so persisting this information is a good idea.